### ISSUE TITLE
publish snapshot as part of the regular build pipeline

### ISSUE THREAD

**Issue by SylvainJuge** [elastic/apm-agent-java#1056]
# What we currently have

- one scheduled job on `internal-ci` that rebuilds the whole agent from `master` with the `@daily` schedule and publishes the snapshot to sonatype maven repository takes (40 to 60 minutes).
- one build pipeline that rebuilds the agent on every update of `master` (and also other branches).

# Improvement proposal

- add a step to the pipeline that will publish last built (and tested) snapshot to sonatype repository
- make this step only executed on the `master` branch.
- remove the scheduled job to publish snapshot

This is independant (but somehow related) to the ongoing effort to automate release process https://github.com/elastic/apm-agent-java/issues/991

# Benefits

- one less build job to take care of
- everything is described in pipeline
- avoid testing and rebuilding

**Comment by cachedout** [comment]
> add a step to the pipeline that will publish last built (and tested) snapshot to sonatype repository

I assume here that you mean [this pipeline ](https://github.com/elastic/apm-agent-java/blob/master/Jenkinsfile) which runs on the [public-facing Jenkins instance](https://apm-ci.elastic.co).

We can do this, but in the past there have been some reservations expressed by @felixbarny and perhaps others about interacting with Nexus via the public Jenkins, and the risk of potential leaks of credentials, etc.

Before I start this work, I want to make sure we're all comfortable with the public-facing CI doing this work. Otherwise, we can take another look at doing this on the internal CI or find another strategy.

cc: @felixbarny and @eyalkoren

**Comment by felixbarny** [comment]
IMO, the snapshot build should run on the same Jenkins with a very similar configuration than the release job as it's a good proxy for whether the release job might fail.

But as we have removed the testing step on the release job (we instead require the master to be green) we could do something similar for the snapshot job. Or even trigger the snapshot job after a successful build on master.

**Comment by cachedout** [comment]
> IMO, the snapshot build should run on the same Jenkins with a very similar configuration than the release job as it's a good proxy for whether the release job might fail.

If we keep this on the internal CI, I see the following possible changes available to us:

1) Currently, the snapshot job on the internal CI runs by using the `deploy` task in Maven run directly from a shell. We could replace this by extending the existing Nexus pipeline code to cover the use case of uploading snapshots, as it does not currently do this. If we do this in order to have a close analog to the code which has been written to integrate with the Nexus repo for the release use case, I am not sure that we will have enough overlap in the code paths to make uploading snapshots a good proxy for what needs to be done during the release process, which involves a fair deal more complexity. I'm certainly open to this change, but would like to discuss the pros and cons a bit more before committing to it, though.

2) We could modify the snapshot job to trigger directly on changes to master instead of simply running once per hour. I would like to understand if there is a need for the snapshots to be uploaded immediately after a change, however, or if a delay is acceptable. If we can articulate gains made by doing this, we can definitively make this change.

Apologies for the continual questions but I just want to make sure we're making the right decisions. Thanks as always. Let me know what you think, @SylvainJuge and @felixbarny

**Comment by SylvainJuge** [comment]
Do we have any way to trigger jobs between internal and public CIs ?

I understand the requirement to use internal-ci when using sensitive credentials, but I'm wondering if we could just 'delegate' a few tasks from the public CI to the internal one ?

For example, given this requirement of using `internal-ci`, does something like this is possible:
- instead of running `mvn deploy` to publish the snapshot on public CI, we just delegate to internal CI, with git ID as parameter
- create a parameterized job on internal CI that will execute `mvn deploy -DskipTests` on the provided git commit.

If such direct delegation is not possible, I think that having a tag/branch on git that is updated whenever the build on master completes on public CI would help, then the second job will just be triggered when the reference is updated.

**Comment by cachedout** [comment]
> Do we have any way to trigger jobs between internal and public CIs ?

This may be possible but I will need to defer to @ivanf and @v1v for historical precedent and guidance here.

**Comment by cachedout** [comment]
> If such direct delegation is not possible, I think that having a tag/branch on git that is updated whenever the build on master completes on public CI would help, then the second job will just be triggered when the reference is updated.

@SylvainJuge We could also just set up the internal Jenkins to poll for changes on the master branch and kick off a job when they are detected. Let me know if that would do what you're looking for.

**Comment by SylvainJuge** [comment]
The issue with the 'poll for changes on master branch' strategy is that it will just do the same as the current daily job, but when the `master` branch is updated. While this avoids waiting for up to a day until snapshot is published, it does not really avoid re-testing everything that has already been done by the public CI.

Our build tends to take a while (up to an hour), and includes lots of integration tests that tend to be flaky, thus currently any random failure on master will make us re-launch both jobs, triage build failure notifications, and in the end we still have to wait a lot until the snapshot is available.

Thus, without using any fancy way to delegate job execution between the two CI systems, my proposal would be the following:
- public CI builds on `master` whenever there are new commits, as it currently do with the main build pipeline
- public CI updates a branch `snapshot-master` after a successful build and test execution (should force update if needed)
- internal CI job is triggered on changes on the `snapshot-master` branch, and only does the snapshot deployment `mvn deploy -DskipTests`.

If we can allow the public CI to update some git branches, that should work.

While the binary is still built twice, tests aren't duplicated and we get a snapshot published whenever the build on `master` is green.

Then, we could also apply this to the release job, which could in the end enable having only a single pipeline on public CI that will delegate some sensitive tasks to internal CI.

**Comment by cachedout** [comment]
OK, I think I see more of what you're going for.

What about something like this?

We already have build artifacts for the master branch on the public CI. [See an example of those here](https://apm-ci.elastic.co/job/apm-agent-java/job/apm-agent-java-mbp/job/master/292/artifact/src/github.com/elastic/apm-agent-java/elastic-apm-agent/target/).

So, using the `lastSuccessfulBuild` target we should be able to construct a URL for the most recent successful build artifacts on master that looks like this: https://apm-ci.elastic.co/job/apm-agent-java/job/apm-agent-java-mbp/job/master/lastSuccessfulBuild/artifact/src/github.com/elastic/apm-agent-java/elastic-apm-agent/target/elastic-apm-agent-1.14.1-SNAPSHOT.jar

Using that artifact, could we just pull it down and then deploy it from the internal CI?

**Comment by SylvainJuge** [comment]
Yes, reusing the built artifact from public CI might be an option.

- how would you trigger the deployment on internal CI after a successful build on public CI ?
- you'll have to re-implement what `mvn deploy` does, probably with a `curl` command, for example dealing with authentication details or uploading `pom.xml` to the remote repository. If you want to reuse mvn tasks using a checkout of the project and just put the snapshot JAR in the expected location, you'll likely have previous steps in the build lifecycle executed (compile classes, tests, ...), thus we don't really gain much from reusing the binary.

Thus, apart from avoiding re-building the binary (which isn't the main), sticking to `mvn deploy` might be easier to implement reusing what we currently have.

**Comment by cachedout** [comment]
I put up [this PR ](https://github.com/elastic/apm-agent-java/pull/1096) which may provide a path forward. We can use this on the internal CI to periodically poll the public CI for the most recent passing build on the master branch. If we find a new commit hash respective to the last time we checked, we can assume that we are dealing with a fresh change and then proceed with the snapshot build (without tests) on the internal CI.

@SylvainJuge Let me know what you think. Thanks!

**Comment by cachedout** [comment]
I have [put up a new PR](https://github.com/elastic/apm-agent-java/pull/1167) that just deploys snapshots from the master branch in the public CI. I think this is the easiest and most straightforward way and it aligns with the original proposal from @SylvainJuge

**Comment by cachedout** [comment]
@SylvainJuge Given the discussion in https://github.com/elastic/apm-agent-java/pull/1167 do you think this issue should remain open?
